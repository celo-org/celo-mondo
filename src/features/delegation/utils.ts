import { accountsABI } from '@celo/abis';
import { App, Octokit } from 'octokit';
import path from 'path';
import { fornoRpcUrl } from 'src/config/config';
import { Addresses } from 'src/config/contracts';
import {
  delegateeRegistrationRequestToMetadata,
  metadataToJSONString,
} from 'src/features/delegation/delegateeMetadata';
import { RegisterDelegateRequest } from 'src/features/delegation/types';
import { createPublicClient, http } from 'viem';
import { celo } from 'viem/chains';

export async function isAddressAnAccount(address: HexString) {
  const client = createPublicClient({ chain: celo, transport: http(fornoRpcUrl) });

  return await client.readContract({
    address: Addresses.Accounts,
    abi: accountsABI,
    functionName: 'isAccount',
    args: [address],
  });
}

const fetchFileShaIfExists = async (
  octokit: Octokit,
  repoOwner: string,
  repoName: string,
  branchRef: string,
  filePath: string,
): Promise<string | null> => {
  try {
    const file = await octokit.rest.repos.getContent({
      owner: repoOwner,
      repo: repoName,
      path: filePath,
      ref: branchRef,
    });

    return file.data.sha;
  } catch (err) {
    const notFound = err instanceof Error && 'status' in err && err.status === 404;

    if (!notFound) {
      throw err;
    }
  }

  return null;
};

export async function createDelegationPR(request: RegisterDelegateRequest) {
  const GITHUB_APP_ID = process.env['GITHUB_APP_ID'] as string;
  const GITHUB_PRIVATE_KEY = process.env['GITHUB_PRIVATE_KEY'] as string;
  const GITHUB_INSTALLATION_ID = parseInt(process.env['GITHUB_INSTALLATION_ID'] as string);

  const app = new App({
    appId: GITHUB_APP_ID,
    privateKey: GITHUB_PRIVATE_KEY,
  });

  const GITHUB_REPO_OWNER = process.env['GITHUB_REPO_OWNER'] as string;
  const GITHUB_REPO_NAME = process.env['GITHUB_REPO_NAME'] as string;
  const octokit = await app.getInstallationOctokit(GITHUB_INSTALLATION_ID);

  const mainBranch = await octokit.rest.repos.getBranch({
    owner: GITHUB_REPO_OWNER,
    repo: GITHUB_REPO_NAME,
    branch: 'main',
  });

  const branchName = `delegatee/${request.address}-${Date.now()}`;
  const metadataPath = `src/config/delegatees/${request.address}.json`;
  const imagePath = `public/logos/delegatees/${request.address}${path.extname(request.image!.name)}`;
  const delegateeMetadata = delegateeRegistrationRequestToMetadata(request, new Date());

  const newRefResponse = await octokit.rest.git.createRef({
    owner: GITHUB_REPO_OWNER,
    repo: GITHUB_REPO_NAME,
    ref: `refs/heads/${branchName}`,
    sha: mainBranch.data.commit.sha,
  });

  const [metadataFileSha, imageFileSha] = await Promise.all([
    fetchFileShaIfExists(
      octokit,
      GITHUB_REPO_OWNER,
      GITHUB_REPO_NAME,
      newRefResponse.data.ref,
      metadataPath,
    ),
    fetchFileShaIfExists(
      octokit,
      GITHUB_REPO_OWNER,
      GITHUB_REPO_NAME,
      newRefResponse.data.ref,
      imagePath,
    ),
  ]);

  await Promise.all([
    octokit.rest.repos.createOrUpdateFileContents({
      owner: GITHUB_REPO_OWNER,
      repo: GITHUB_REPO_NAME,
      path: metadataPath,
      message: `Adding delegatee ${request.address} metadata`,
      content: Buffer.from(metadataToJSONString(delegateeMetadata)).toString('base64'),
      sha: metadataFileSha || undefined,
      branch: branchName,
    }),
    octokit.rest.repos.createOrUpdateFileContents({
      owner: GITHUB_REPO_OWNER,
      repo: GITHUB_REPO_NAME,
      path: imagePath,
      message: `Adding delegatee ${request.address} image`,
      content: Buffer.from(await request.image!.arrayBuffer()).toString('base64'),
      sha: imageFileSha || undefined,
      branch: branchName,
    }),
  ]);

  const createPRResponse = await octokit.rest.pulls.create({
    owner: GITHUB_REPO_OWNER,
    repo: GITHUB_REPO_NAME,
    head: branchName,
    base: 'main',
    title: `Adding delegate ${request.name}`,
    body: `Adding delegate \`${request.name}\` with address [${request.address}](https://celoscan.io/address/${request.address}).
      
[Verification](${request.verificationUrl}) (\`${request.verificationUrl}\`)`,
  });

  return createPRResponse.data.html_url;
}
