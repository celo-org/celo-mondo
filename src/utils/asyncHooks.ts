import { useCallback, useEffect, useLayoutEffect, useRef } from 'react';

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect;

// https://usehooks-typescript.com/react-hook/use-interval
export function useInterval(callback: () => void, delay: number | null) {
  const savedCallback = useRef(callback);

  // Remember the latest callback if it changes.
  useIsomorphicLayoutEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  // Set up the interval.
  useEffect(() => {
    // Don't schedule if no delay is specified.
    // Note: 0 is a valid value for delay.
    if (!delay && delay !== 0) {
      return;
    }

    const id = setInterval(() => savedCallback.current(), delay);

    return () => clearInterval(id);
  }, [delay]);
}

// https://medium.com/javascript-in-plain-english/usetimeout-react-hook-3cc58b94af1f
export const useTimeout = (
  callback: () => void,
  delay = 0, // in ms (default: immediately put into JS Event Queue)
): (() => void) => {
  const timeoutIdRef = useRef<any>(null);

  const cancel = useCallback(() => {
    const timeoutId = timeoutIdRef.current;
    if (timeoutId) {
      timeoutIdRef.current = undefined;
      clearTimeout(timeoutId);
    }
  }, [timeoutIdRef]);

  useEffect(() => {
    if (delay >= 0) {
      timeoutIdRef.current = setTimeout(callback, delay);
    }
    return cancel;
  }, [callback, delay, cancel]);

  return cancel;
};
