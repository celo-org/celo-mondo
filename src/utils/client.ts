import { config, fornoRpcUrl } from 'src/config/config';
import { Account, createPublicClient, createWalletClient, http } from 'viem';

export const createCeloPublicClient = () =>
  createPublicClient({
    chain: config.chain,
    transport: config.chain.testnet ? http() : http(fornoRpcUrl),
    batch: {
      multicall: true,
    },
  });

export const createCeloWalletClient = (account: Account) =>
  createWalletClient({
    account,
    chain: config.chain,
    transport: config.chain.testnet ? http() : http(fornoRpcUrl),
  });
