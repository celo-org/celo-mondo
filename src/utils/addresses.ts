import { logger } from 'src/utils/logger';
import { getAddress, isAddress } from 'viem';

export const ADDRESS_REGEX = /^0x[a-fA-F0-9]{40}$/;
export const TX_HASH_REGEX = /^0x[a-fA-F0-9]{64}$/;

export function isValidAddress(address: string): address is Address {
  // Need to catch because ethers' isAddress throws in some cases (bad checksum)
  try {
    const isValid = address && isAddress(address);
    return !!isValid;
  } catch (error) {
    logger.warn('Invalid address', error, address);
    return false;
  }
}

export function validateAddress(address: string, context: string) {
  if (!address || !isAddress(address)) {
    const errorMsg = `Invalid addresses for ${context}: ${address}`;
    logger.error(errorMsg);
    throw new Error(errorMsg);
  }
}

export function normalizeAddress(address: string) {
  return getAddress(address);
}

export function shortenAddress(address: string, capitalize = true) {
  validateAddress(address, 'shorten');
  const normalizedAddress = normalizeAddress(address);

  const start = normalizedAddress.substring(0, 6);
  const end = normalizedAddress.substring(address.length - 4);

  const shortened = `${start}â€¦${end}`;
  return capitalize ? capitalizeAddress(shortened) : shortened;
}

export function capitalizeAddress(address: string) {
  return '0x' + address.substring(2).toUpperCase();
}

export function eqAddress(a1: string, a2: string) {
  return normalizeAddress(a1) === normalizeAddress(a2);
}

export function eqAddressSafe(a1: string, a2: string) {
  try {
    return eqAddress(a1, a2);
  } catch (error) {
    logger.error('Error comparing addresses', error, a1, a2);
    return false;
  }
}

export function ensure0x(hexstr: string) {
  return hexstr.startsWith('0x') ? hexstr : `0x${hexstr}`;
}

export function strip0x(hexstr: string) {
  return hexstr.startsWith('0x') ? hexstr.slice(2) : hexstr;
}
