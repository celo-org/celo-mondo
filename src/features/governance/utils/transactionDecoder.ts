import { governanceABI } from '@celo/abis';
import { Addresses } from 'src/config/contracts';
import { celoPublicClient } from 'src/utils/client';
import { Address, decodeFunctionData, formatUnits } from 'viem';

export interface DecodedTransaction {
  to: Address;
  value: string;
  data: string;
  functionName: string;
  args: Record<string, any>;
  description: string | undefined;
  contractName: string;
}

export interface ProposalTransaction {
  index: number;
  to: Address;
  value: bigint;
  data: `0x${string}`;
  decoded?: DecodedTransaction;
  error?: string;
}

/**
 * Fetches all transactions for a given proposal
 */
export async function getProposalTransactions(
  proposalId: number,
  transactionCount: number,
  blockNumber: bigint,
): Promise<ProposalTransaction[]> {
  try {
    if (transactionCount === 0) {
      return [];
    }

    const baseCall = {
      abi: governanceABI,
      address: Addresses.Governance,
      functionName: 'getProposalTransaction',
    } as const;

    const transactions = await celoPublicClient.multicall({
      allowFailure: false,
      blockNumber,
      contracts: Array(transactionCount)
        .fill(1)
        .map((_, index) => ({
          ...baseCall,
          args: [BigInt(proposalId), BigInt(index)],
        })),
    });

    return transactions.map(([value, to, data], index) => ({
      value,
      to,
      data,
      index,
    }));
  } catch (error) {
    console.error(`Error fetching transactions for proposal ${proposalId}@${blockNumber}:`, error);
    return [];
  }
}

/**
 * Decodes a single transaction using common contract ABIs
 */
export async function decodeTransaction(
  transaction: ProposalTransaction,
): Promise<DecodedTransaction | null> {
  if (!transaction.data || transaction.data === '0x') {
    return null;
  }

  try {
    // Try to decode with governance ABI first
    try {
      const decoded = decodeFunctionData({
        abi: governanceABI,
        data: transaction.data,
      });

      return {
        to: transaction.to,
        value: formatUnits(transaction.value, 18),
        data: transaction.data,
        functionName: decoded.functionName,
        args: decoded.args as Record<string, any>,
        description: getTransactionDescription(
          decoded.functionName,
          decoded.args as unknown as any[],
        ),
        contractName: 'Governance',
      };
    } catch {
      // If governance ABI fails, try with common function selectors
      return await decodeWithCommonSelectors(transaction);
    }
  } catch (error) {
    console.error('Error decoding transaction:', error);
    return null;
  }
}

/**
 * Decodes transactions using common function selectors when ABI is not available
 */

// Common function selectors
const commonSelectors = import('src/config/selectors.json');
const BASE_URL_4BYTE = `https://www.4byte.directory/api/v1/signatures`;

async function decodeWithCommonSelectors(
  transaction: ProposalTransaction,
): Promise<DecodedTransaction | null> {
  const selector = transaction.data.slice(0, 10);

  const selectors = (await commonSelectors).default;
  let selectorInfo: string | undefined = selectors[selector as keyof typeof selectors];
  if (!selectorInfo) {
    const sig: {
      count: number;
      next: number;
      previous: number;
      results: [
        {
          id: number;
          created_at: string;
          text_signature: string;
          hex_signature: `0x${string}`;
          bytes_signature: string;
        },
      ];
    } = await fetch(`${BASE_URL_4BYTE}/?hex_signature=${selector}`).then((x) => x.json());

    if (sig.count > 0) {
      const functionName = sig.results.at(-1)!.text_signature;
      // @ts-expect-error
      selectors[selector] = functionName;
      selectorInfo = functionName;
    }
  }

  if (selectorInfo) {
    const functionName = selectorInfo.split('(')[0];
    const inputTypes = (selectorInfo as string).match(/\((.+)\)/)![1].split(',');
    const decoded = decodeFunctionData({
      abi: [
        {
          type: 'function',
          name: functionName,
          inputs: inputTypes.map((type, index) => ({ name: index.toString(), type })),
        },
      ],
      data: transaction.data,
    });

    return {
      to: transaction.to,
      value: formatUnits(transaction.value, 18),
      data: transaction.data,
      functionName,
      args: decoded.args!.map((input, index) => `(${inputTypes[index]}) ${input}`),
      description: getTransactionDescription(
        decoded.functionName,
        decoded.args as unknown as any[],
      ),
      contractName: getContractName(transaction.to),
    };
  }

  // push to commonSelectors

  return {
    to: transaction.to,
    value: formatUnits(transaction.value, 18),
    data: transaction.data,
    functionName: 'unknown',
    args: {},
    description: undefined,
    contractName: getContractName(transaction.to),
  };
}

function getTransactionDescription(functionName: string, args: any[]) {
  switch (functionName.slice(0, functionName.indexOf('('))) {
    case 'setRegistry':
      return `Set registry to ${args[0]}`;
    case 'setApprover':
      return `Set approver to ${args[0]}`;
    case 'setConcurrentProposals':
      return `Set concurrent proposals to ${args[0]}`;
    case 'setMinDeposit':
      return `Set minimum deposit to ${formatUnits(args[0], 18)} CELO`;
    case 'setQueueExpiry':
      return `Set queue expiry to ${args[0]} seconds`;
    case 'setDequeueFrequency':
      return `Set dequeue frequency to ${args[0]} seconds`;
    case 'setApprovalStageDuration':
      return `Set approval stage duration to ${args[0]} seconds`;
    case 'setReferendumStageDuration':
      return `Set referendum stage duration to ${args[0]} seconds`;
    case 'setExecutionStageDuration':
      return `Set execution stage duration to ${args[0]} seconds`;
    case 'setParticipationBaseline':
      return `Set participation baseline to ${formatUnits(args[0], 18)} CELO`;
    case 'setParticipationFloor':
      return `Set participation floor to ${formatUnits(args[0], 18)} CELO`;
    case 'setBaselineUpdateFactor':
      return `Set baseline update factor to ${formatUnits(args[0], 18)}`;
    case 'setBaselineQuorumFactor':
      return `Set baseline quorum factor to ${formatUnits(args[0], 18)}`;
    default:
      return undefined;
  }
}

const CONTRACTS: Record<string, string> = {
  // CORE CONTRACTS
  '0x9d65e69ac940dcb469fd7c46368c1e094250a400': 'Reserve',
  '0x7d21685c17607338b313a7174bab6620bad0aab7': 'Accounts',
  '0xdc553892cdeeed9f575aa0fba099e5847fd88d20': 'Attestations',
  '0x0ad5b1d0c25ecf6266dd951403723b2687d6aff2': 'FederatedAttestations',
  '0xae6b29f31b96e61dddc792f45fda4e4f0356d0cb': 'OdisPayments',
  '0x6e10a8864c65434a721d82e424d727326f9d5bfa': 'BlockchainParameters',
  '0x7a8c7a833565fc428cdfba20fe03fafb178a434f': 'CeloUnreleasedTreasury',
  '0x50c100bacde7e2b546371eb0be1eaccf0a6772ec': 'DoubleSigningSlasher',
  '0x71cac3b31c138f3327c6ca14f9a1c8d752463fdd': 'DowntimeSlasher',
  '0x8d6677192144292870907e3fa8a5527fe55a7ff6': 'Election',
  '0xf424b5e85b290b66ac20f8a9eab75e25a526725e': 'EpochManager',
  '0x2d4148c3500f696aa5a83dd4cc35c289b738b687': 'EpochManagerEnabler',
  '0x07f007d389883622ef8d4d347b3f78007f28d8b7': 'EpochRewards',
  '0xf4fa51472ca8d72af678975d9f8795a504e7ada5': 'Escrow',
  '0x67316300f17f063085ca8bca4bd3f7a5a3c66275': 'Exchange',
  '0xe383394b913d7302c49f794c7d3243c429d53d1d': 'ExchangeEUR',
  '0x15f344b9e6c3cb6f0376a36a64928b13f62c6276': 'FeeCurrencyDirectory',
  '0xbb024e9cdcb2f9e34d893630d19611b8a5381b3c': 'FeeCurrencyWhitelist',
  '0x47a472f45057a9d79d62c6427367016409f4ff5a': 'Freezer',
  '0xcd437749e43a154c07f3553504c68fbfd56b8778': 'FeeHandler',
  '0xdfca3a8d7699d8bafe656823ad60c17cb8270ecc': 'GasPriceMinimum',
  '0x471ece3750da237f93b8e339c536989b8978a438': 'GoldToken (Celo Token)',
  '0xd533ca259b330c7a88f74e000a3faea2d63b7972': 'Governance',
  '0x03f6842b82dd2c9276931a17dd23d73c16454a49': 'GrandaMento',
  '0x6cc083aed9e3ebe302a6336dbc7c921c9f03349e': 'LockedGold (LockedCelo)',
  '0x4efa274b7e33476c961065000d58ee09f7921a74': 'MentoFeeHandlerSeller',
  '0x22a4aaf42a50bfa7238182460e32f15859c93dfe': 'Random',
  '0x000000000000000000000000000000000000ce10': 'Registry',
  '0x9380fa34fd9e4fd14c06305fd7b6199089ed4eb9': 'Reserve',
  '0xef3b9cc0fa4717af6f412d39dbceb89bf92f603b': 'ScoreManager',
  '0xefb84935239dacdecf7c5ba76d8de40b077b7b33': 'SortedOracles',
  '0xb49e4d6f0b7f8d0440f75697e6c8b37e09178bcf': 'TransferWhitelist',
  '0xd3aee28548dbb65df03981f0dc0713bfcbd10a97': 'UniswapFeeHandlerSeller',
  '0xaeb865bca93ddc8f47b8e29f40c5399ce34d0c58': 'Validators',

  // WHITELISTED TOKENS CONTRACTS
  '0x48065fbbe25f71c9282ddf5e1cd6d6a887483d5e': 'Tether USD',
  '0x0e2a3e05bc9a16f5292a6170456a710cb89c6f72': 'Tether USD',
  '0x105d4a9306d2e55a71d2eb95b81553ae1dc20d7b': 'PUSO',
  '0xceba9300f2b948710d2653dd7b07f33a8b32118c': 'USDC',
  '0x2f25deb3848c207fc8e0c34035b3ba7fc157602b': 'USDC',
  '0x456a3d042c0dbd3db53d5489e98dfb038553b0d0': 'Celo Kenyan Shilling',
  '0x4c35853a3b4e647fd266f4de678dcc8fec410bf6': 'Celo South African Rand',
  '0x7175504c455076f15c04a2f90a8e352281f492f9': 'Celo Australian Dollar',
  '0x73f93dcc49cb8a239e2032663e9475dd5ef29a08': 'ECO CFA',
  '0x765de816845861e75a25fca122bb6898b8b1282a': 'Celo Dollar',
  '0x8a567e2ae79ca692bd748ab832081c45de4041ea': 'Celo Colombian Peso',
  '0xccf663b1ff11028f0b19058d0f7b674004a40746': 'Celo British Pound',
  '0xd8763cba276a3738e6de85b4b3bf5fded6d6ca73': 'Celo Euro',
  '0xe2702bd97ee33c88c8f6f92da3b733608aa76f71': 'Celo Nigerian Naira',
  '0xb55a79f398e759e43c95b979163f30ec87ee131d': 'Celo Swiss Franc',
  '0xc45ecf20f3cd864b32d9794d6f76814ae8892e20': 'Celo Japanese Yen',
  '0xe8537a3d056da446677b9e9d6c5db704eaab4787': 'Celo Brazilian Real',
  '0xfaea5f3404bba20d3cc2f8c4b0a888f55a3c7313': 'Celo Ghanaian Cedi',
  '0xff4ab19391af240c311c54200a492233052b6325': 'Celo Canadian Dollar',
};
export function getContractName(address: Address): string {
  if (!address) {
    return 'Unknown Contract';
  }

  return CONTRACTS[address.toLowerCase()] || 'Unknown Contract';
}
