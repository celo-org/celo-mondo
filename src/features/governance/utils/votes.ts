import { governanceABI } from '@celo/abis';
import { Event, fetchProposalEvents } from 'src/app/governance/events';
import { EmptyVoteAmounts, VoteAmounts, VoteType } from 'src/features/governance/types';
import { isValidAddress } from 'src/utils/addresses';
import { celoPublicClient } from 'src/utils/client';
import { logger } from 'src/utils/logger';
import { bigIntSum } from 'src/utils/math';
import { objFilter } from 'src/utils/objects';
import { decodeEventLog } from 'viem';

export async function sumProposalVotes(id: number): Promise<{
  voters: AddressTo<VoteAmounts>;
  totals: VoteAmounts;
}> {
  const castVoteEvents = await fetchProposalEvents(celoPublicClient.chain.id, 'ProposalVoted', {
    proposalId: BigInt(id),
  });

  const voterToVotes: AddressTo<VoteAmounts> = {};
  reduceLogs(voterToVotes, castVoteEvents, true);

  // Skipping revoke vote events for now for performance since they are almost never used
  // Much more common is for a voter to re-vote, replacing the old vote
  // const revokeVoteParams = `topic0=${revokeVoteTopics[0]}&topic1=${revokeVoteTopics[1]}&topic0_1_opr=and`;
  // const revokeVoteEvents = await queryCeloscanLogs(Addresses.Governance, revokeVoteParams);
  // reduceLogs(voterToVotes, revokeVoteEvents, false);

  // Filter out voters with no current votes
  const voters = objFilter(
    voterToVotes,
    (_, votes): votes is VoteAmounts => bigIntSum(Object.values(votes)) > 0n,
  );

  const totals = Object.values(voters).reduce<VoteAmounts>(
    (acc, votes) => {
      acc[VoteType.Yes] += votes[VoteType.Yes];
      acc[VoteType.No] += votes[VoteType.No];
      acc[VoteType.Abstain] += votes[VoteType.Abstain];
      return acc;
    },
    { ...EmptyVoteAmounts },
  );

  return { voters, totals };
}

function reduceLogs(voterToVotes: AddressTo<VoteAmounts>, logs: Event[], isCast: boolean) {
  for (const log of logs) {
    const decoded = decodeVoteEventLog(log);
    if (!decoded) continue;
    const { account, yesVotes, noVotes, abstainVotes } = decoded;
    if (isCast) {
      voterToVotes[account] = {
        [VoteType.Yes]: yesVotes,
        [VoteType.No]: noVotes,
        [VoteType.Abstain]: abstainVotes,
      };
    } else {
      voterToVotes[account] = EmptyVoteAmounts;
    }
  }
}

export function getLargestVoteType(votes: VoteAmounts) {
  let maxType = VoteType.None;
  let maxValue = 0n;

  for (const [type, value] of Object.entries(votes)) {
    if (value && value > maxValue) {
      maxType = type as VoteType;
      maxValue = value;
    }
  }

  return { type: maxType, value: maxValue };
}

export function decodeVoteEventLog(log: Event) {
  try {
    if (!log.topics || log.topics.length < 3) return null;
    const { eventName, args } = decodeEventLog({
      abi: governanceABI,
      data: log.data,
      topics: log.topics,
      strict: false,
    });

    let proposalId: number | undefined;
    let account: string | undefined;
    let yesVotes: bigint = 0n;
    let noVotes: bigint = 0n;
    let abstainVotes: bigint = 0n;

    if (eventName === 'ProposalVoted' || eventName === 'ProposalVoteRevoked') {
      proposalId = Number(args.proposalId);
      account = args.account;
      yesVotes = (BigInt(args.value!) === 3n && BigInt(args.weight!)) || 0n;
      noVotes = (BigInt(args.value!) === 2n && BigInt(args.weight!)) || 0n;
      abstainVotes = (BigInt(args.value!) === 1n && BigInt(args.weight!)) || 0n;
    } else if (eventName === 'ProposalVotedV2' || eventName === 'ProposalVoteRevokedV2') {
      proposalId = Number(args.proposalId);
      account = args.account;
      yesVotes = BigInt(args.yesVotes!) || 0n;
      noVotes = BigInt(args.noVotes!) || 0n;
      abstainVotes = BigInt(args.abstainVotes!) || 0n;
    } else {
      logger.warn('Invalid event name, expected ProposalVoted', eventName);
      return null;
    }

    if (!proposalId || !account || !isValidAddress(account)) return null;

    return {
      proposalId,
      account,
      yesVotes,
      noVotes,
      abstainVotes,
    };
  } catch (error) {
    logger.warn('Error decoding event log', error, log);
    return null;
  }
}

export function assertEvent<T>(eventNames: readonly T[], eventName: any): eventName is T {
  return eventNames.includes(eventName);
}
