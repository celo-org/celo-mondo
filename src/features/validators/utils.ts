import { ValidatorGroup, ValidatorStatus } from 'src/features/validators/types';
import { eqAddressSafe } from 'src/utils/addresses';
import { toTitleCase, trimToLength } from 'src/utils/strings';

export function findGroup(groups?: ValidatorGroup[], address?: Address) {
  if (!groups || !address) return undefined;
  return groups.find((g) => eqAddressSafe(g.address, address));
}

export function cleanGroupName(name: string) {
  return trimToLength(toTitleCase(name.replace(/group|Group/g, '').replace(/[-_]/g, ' ')), 20);
}

export function isElected(group: ValidatorGroup) {
  return Object.values(group.members).some((m) => m.status === ValidatorStatus.Elected);
}

export function getGroupStats(group?: ValidatorGroup) {
  if (!group) return { numMembers: 0, numElected: 0, score: 0 };
  const members = Object.values(group.members);
  const electedMembers = members.filter((m) => m.status === ValidatorStatus.Elected);
  return { numMembers: members.length, numElected: electedMembers.length, score: group.score };
}
