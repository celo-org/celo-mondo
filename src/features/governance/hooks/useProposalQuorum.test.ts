import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import BigNumber from 'bignumber.js';
import { toFixidity } from 'src/utils/numbers';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { publicArchiveClient, renderHook } from '../../../test/anvil/utils';
import {
  extractFunctionSignature,
  fetchThresholds,
  useParticipationParameters,
  useProposalQuorum,
} from './useProposalQuorum';

beforeEach(() => {
  vi.mock('wagmi', async (importActual) => ({
    ...(await importActual()),
    usePublicClient: () => publicArchiveClient,
    useReadContract: vi.fn(() => ({
      // Mocks the following:
      // abi: governanceABI,
      // functionName: 'getParticipationParameters',
      data: [
        60000000000000000000000n, // baseline
        50000000000000000000000n, // baselineFloor
        200000000000000000000000n, // baselineUpdateFactor
        500000000000000000000000n, // baselineQuorumFactor
      ],
    })),
  }));
});

afterEach(() => {
  vi.clearAllMocks();
});

describe('extractFunctionSignature', () => {
  it('gets the four first bytes of a byte array', () => {
    expect(extractFunctionSignature('0x112233445566778899')).toBe('0x11223344');
    expect(extractFunctionSignature('1234567890' as `0x${string}`)).toBe('0x12345678');
  });
});

describe('fetchThresholds', () => {
  // Check for data sanity at known blocks for known proposals
  // in this case the proposal#195 at block 0x1bb55ea
  it('fetches the constitution threshholds properly #1', async () => {
    // Proposal for Creation of Celo Governance Guild's execution block - 1
    process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER = '0x1bb55ea';
    const expectedNumberOfTxs = 1n;
    const proposalId = 195;
    const spy = vi.spyOn(publicArchiveClient, 'multicall');
    await expect(fetchThresholds(publicArchiveClient, proposalId, expectedNumberOfTxs)).resolves
      .toMatchInlineSnapshot(`
      [
        0.6,
      ]
    `);
    await expect(spy.mock.results[0].value).resolves.toHaveLength(Number(expectedNumberOfTxs));
    expect(spy.mock.calls.length).toBe(2);
  });

  // Check for data sanity at known blocks for known proposals
  // in this case the proposal#16 at block 0x1baa98c
  it('fetches the constitution threshholds properly #2', async () => {
    // Enabling MENTO Governance's proposal block
    process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER = '0x1baa98c';
    const expectedNumberOfTxs = 50n;
    const proposalId = 196;
    const spy = vi.spyOn(publicArchiveClient, 'multicall');
    await expect(
      fetchThresholds(publicArchiveClient, proposalId, expectedNumberOfTxs),
    ).resolves.toMatchSnapshot();
    await expect(spy.mock.results[0].value).resolves.toHaveLength(Number(expectedNumberOfTxs));
    expect(spy.mock.calls.length).toBe(2);
  });
});

describe('useParticipationParameters', () => {
  it('fetches the params from the governance contract', async () => {
    const { result } = renderHook(() => useParticipationParameters());
    await waitFor(() => expect(result.current.isLoading).not.toBe(true));
    expect(result.current).toMatchInlineSnapshot(`
      {
        "data": {
          "baseline": 0.06,
          "baselineFloor": 0.05,
          "baselineQuorumFactor": 0.5,
          "baselineUpdateFactor": 0.2,
        },
        "error": undefined,
        "isLoading": undefined,
      }
    `);
  });
});

describe('useProposalQuorum', () => {
  /*
    This test makes sure the calculation is correct
    following these two solidity functions:
    - https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Governance.sol#L1567
    - https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/governance/Proposals.sol#L195-L211

    Let:
    - baseline and baselineQuorumFactor are part of the `getParticipationParameters` results
    - quorumVotes = proposal.networkWeight * baseline * baselineQuorumFactor
    - thresholds = the array of `getConstitutionContract` results corresponding of each transaction of a proposal
    - maxThreshold = the highest thresholds in the above array

    The quorum to "beat" should be: 
    - maxThreshold * quorumVotes where
   */
  it('maths properly', async () => {
    // Enabling MENTO Governance's proposal block
    process.env.NEXT_PUBLIC_FORK_BLOCK_NUMBER = '0x1baa98c';
    const baseline = 0.06;
    const baselineQuorumFactor = 0.5;
    let calls = 0;
    vi.spyOn(publicArchiveClient, 'multicall').mockImplementation(async () => {
      if (calls++ === 0) {
        return [
          ['0x', '0x', '0x'],
          ['0x', '0x', '0x'],
          ['0x', '0x', '0x'],
        ];
      } else {
        return [0.75, 0.1, 0.5].map(toFixidity);
      }
    });
    const networkWeight = 200000000000000000000000000n;

    const quorumPct = baseline * baselineQuorumFactor;
    const quorumVotes = new BigNumber(networkWeight.toString()).times(quorumPct);
    const maxThreshold = quorumVotes.times(0.75 /* highest mocked threshold */);

    const { result } = renderHook(() =>
      // @ts-expect-error
      useProposalQuorum({ proposal: { id: 1, numTransactions: 3, networkWeight } }),
    );
    await waitFor(() => expect(result.current.isLoading).not.toBe(true));
    expect(result.current.data).toBe(BigInt(maxThreshold.toFixed(0)));
    expect(BigInt(maxThreshold.toFixed(0))).toMatchInlineSnapshot(`4500000000000000000000000n`);
  });
});
