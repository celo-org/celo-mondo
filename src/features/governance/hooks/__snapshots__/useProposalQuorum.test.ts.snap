// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchThresholds > fetches the constitution threshholds properly #2 1`] = `
[
  0.5,
  0.5,
  0.5,
  0.9,
  0.9,
  0.9,
  0.9,
  0.5,
  0.8,
  0.9,
  0.5,
  0.9,
  0.9,
  0.9,
  0.8,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.9,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.8,
  0.9,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
  0.5,
]
`;
